# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
import os
os.listdir()

['9_5_Класи.ipynb',
 'hw_10_2_Основи_роботи_з_даними_у_Pandas.ipynb',
 'hw_9_1_Операції_над_даними.ipynb',
 'hw_9_3_Цикл_while.ipynb',
 'hw_9_3_Comprehensions.ipynb',
 'hw_9_2_Control_Flow.ipynb',
 'hw_9_2_Списки.ipynb',
 'hw_9_2_Словники,_набори,_кортежі.ipynb',
 'hw_9_1_Cинтаксис_Python (1).ipynb',
 'hw_9_3_Цикл_for.ipynb',
 '9_4_Функції_Продовження.ipynb',
 'ДЗ Модуль 3.1.docx',
 '9_5_Модулі_Винятки.ipynb',
 '9_4_Функціїї_Основи.ipynb',
 'hw_10_1_Введення_у_Pandas_та_NumPy_.ipynb',
 '.ipynb_checkpoints',
 'hw_9_1_Cинтаксис_Python (2).ipynb',
 'hw_9_1_Cинтаксис_Python.ipynb']

In [3]:
os.listdir('../data')

['.DS_Store',
 'archive',
 'sample_superstore.xls',
 'supermarket_sales.csv',
 'archive.zip']

In [4]:
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [5]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [6]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [7]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [8]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [9]:
df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

На 75й персентиль припадає значення ~ 22,45 дол. Це означає, що 3/4 всіх покупок, здійснених в супермаркетах були від межах від 0,51 (мінімум) до 22,45 дол.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [10]:
df.City.value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [11]:
df.City.value_counts(normalize=True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [12]:
df.rename(columns={'gross income' : 'gross_income'}, inplace=True)

In [24]:
filter_1 = df.City =='Yangon'
yangon_count = df[filter_1].gross_income.count()
filter_1 = (df.City =='Yangon')
yangon_sum = df[filter_1].gross_income.sum()
yangon_avg = yangon_sum / yangon_count
print(f'Yangon gross income = {yangon_sum}\nYangon count = {yangon_count} \nYangon gross income AVG = {yangon_avg}')

Yangon gross income = 5057.1605
Yangon count = 340 
Yangon gross income AVG = 14.874001470588235


In [25]:
filter_2 = df.City =='Mandalay'
mandalay_count = df[filter_2].gross_income.count()
filter_2 = (df.City =='Mandalay')
mandalay_sum = df[filter_2].gross_income.sum()
mandalay_avg = mandalay_sum / mandalay_count
print(f'Mandalay gross income = {mandalay_sum}\nMandalay count = {mandalay_count} \nMandalay gross income AVG = {mandalay_avg}')

Mandalay gross income = 5057.032
Mandalay count = 332 
Mandalay gross income AVG = 15.232024096385542


В Yangon незначно більшим є сукупний gross income, а в Mandalay більшим є gross income AVG. 
Можливі причини більшого загального gross income (в Yangon):
1) більше покупців (або покупок)
2) продано більше товарів на покупця (більшим є середній чек)
3) більше торгових годин в Yangon.
Причини більшого середнього gross income (в Mandalay):
1) в Mandalay продається більше дорожчих категорій товарів
2) в Mandalay менша кількість покупців (покупок) при однаковому рівні виручки
3) в Mandalay gross income AVG = 15.23 більш зручне розташування торгівельних точок, відсутні конкуренти, тому покупці витрачають більше саме в магазинах цієї мережі.
Перевірка гіпотез:
    Про загальний дохід:
1.1. Порахувати yangon_count (340) та mandalay_count (332), порівняти їх. Кількість покупців в Янгоні більша - гіпотеза підтверджена.
1.2. порахувати середній чек для обох локацій: Yangon gross income AVG = 14.87, Mandalay gross income AVG = 15.23. Гіпотеза спростована, середній чек в Янгоні менший.
1.3. порівняти активність покупців за годинами
    Про середній дохід:
2.1. сегментувати продукти за unit price та порівняти за обома містами
2.2. так, кількість покупців менша при однаковій виручці, гіпотеза підтверджена.
2.3. потрібні додаткові дані щодо розташування конкурентів в містах.
   Висновки та рекомендації:
- магазини працюють практично однаково в обох містах, різниця в ключових показниках незначна;
- потрібно отримати перелік продуктів, що продаються найкраще в Mandalay та просувати їх в Янгоні.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [26]:
df.rename(columns={'gross margin percentage' : 'gross_margin_percentage'}, inplace=True)

Перевіримо значення маржі троьома методами:

In [31]:
df.gross_margin_percentage.describe()

count    1.000000e+03
mean     4.761905e+00
std      6.131498e-14
min      4.761905e+00
25%      4.761905e+00
50%      4.761905e+00
75%      4.761905e+00
max      4.761905e+00
Name: gross_margin_percentage, dtype: float64

In [38]:
df.gross_margin_percentage.unique()

array([4.76190476])

In [36]:
df.gross_margin_percentage.nunique()

1

Отже, за методом describe() бачимо, що мінімальне, максимальне, середнє значення та значення 25,50,75-го квартилів співпадають. За методом unique(), отримали одне число у масиві. За методом nunique() підтвердили, що число одне. У всіх всіх інвойсах однакова маржа 4.76190476

Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [48]:
df.rename(columns={'Product line' : 'product_line'}, inplace=True)

In [58]:
filter_3 = df.City =='Yangon'
categoris_yangon = df[filter_3].product_line.value_counts(normalize=True)
categoris_yangon

product_line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: proportion, dtype: float64

In [59]:
filter_4 = df.City =='Mandalay'
categoris_mandalay = df[filter_4].product_line.value_counts(normalize=True)
categoris_mandalay

product_line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: proportion, dtype: float64

Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)